### ISSUE TITLE
Is maintaining and developing OT even worth the effort as a 'getting into animation' software?

### ISSUE THREAD

**Issue by Lovewyrm** [opentoonz/opentoonz#3103]
I've felt pretty good about OT for some time (after all, I was among the ones who said 'no keep OT as its own project when there was a debate about it last year), but lately, these sentiments have soured significantly.

Wouldn't it be better to, seeing how there's so much awful 'old production' cruft in it, to just remake the entire thing instead of working with the old codebase?

It's just so tiring and awkward to use and it has so many troubles, but does it really do much for the average newcomer?
Wouldn't it be better to just use it as a template for a rewrite? 

I do like OT, but I think its being held back by its age and transitional license stage, while still being somewhat entangled with owners who have their own agenda.

Why not take what's good about it and replicate that with a new program?
I'm seeing 344 forks so far, but is forking even the way to go? 

It uses old OpenGL, it uses Qt in a way that doesn't work with Linux and tablets very well.
It uses very specific things most people don't need, etc.

It's very unstable and the issues keep piling up. Why not freeze OT as it is and make something new?
Am I just too disheartened? Maybe. Should this have gone in the google group? Maybe.

It's just a death of a thousand cuts for me. There's always something that sours things up.
Can drag an xsheet column down/longer, by pulling at the tab, but not upwards etc. etc ET CETERA.

Blehhh.
Just close this issue if you believe I'm just a whiner right now. I won't be mad. I'd rather see that than stockholm syndrome excuses.
I'm tired of workarounds for something that is allegedly 'professional grade'.
If the workflow of the studios that used it was so different to the usage that seems to be the norm now (independent artists dabbling with it), then what's the point for us?
Why endure this torture? For a better future? Well, again, what does OT do technically that puts it ahead.

Its opengl code is old, many things are old and awkward. OT is not technically ahead, but it's got a good core for working with animation as a...thing. A suite. 
Just copy that...just copy the cool stuff but don't keep this rancid corpse alive.

**tl; dr:**
What does Joe and Josephine Sixpack gain from this whole thing?
Is OT worth the upkeep versus the tools it provides? Or is it just a 'feels professional, bro' thing?
Would it not be better to just rewrite it, more sensibly this time, and for a general audience?

**Comment by beeheemooth** [comment]
@Lovewyrm You are absolutely right. Probably in the eyes of many developers, OT looks like a sinking ship, which is completely covered with patches. I can even say with confidence that many of them could realize their unique ideas for developing the program if they did not understand that in the current state of affairs this has no future. A tree with rotten roots will fall sooner or later and all the "branches" will disappear.

**Comment by Lovewyrm** [comment]
I think I'd just like to see a more "unixy" approach to the whole art program ecosystem as a whole.
Like the mypaint brushes, which are found pretty much everywhere now.

Have one high speed raster/vector rendering engine, with 'smart' canvases that have metadata (mypaint, again, has this already), because that's where most things hinge on.

Then build things like xsheets and all that around that (and other) components.
Once more, I really like OTs 'spirit', it's 'shell' or whatever you may call it. There is some smart stuff at the core, which makes trying to get at it through the cruft all the more painful.

That said, if there were a rewrite, then I hope it will be highly specialized. No more smooshing bits into other programs. Blender and Krita, for example, both have '2d animation' capabilities.
Blender being a bit more niche in its 3D nature on that regard, but what I'd love to see the most is one super specialized program that plays nice with others.

I don't want a landscape like in text editors...so much crap there, too, while inreality it's basically only three camps.
modal editors like vim, non modal editors like sublime text, notepad, pico etc, and emacs, which is its own thing.

But even if it's just the interface around some core, one thing is common to them all, they need brushes and playing images in sequence.
Just need a really good engine for that, the rest is less of a hassle.

I'm rambling now, though, but yeah, sobering thoughts for me after trying to hold on to 'hey, we have pro stuff in the open source scene now' honeymoonery.

I simply have lost all faith in it being useful for anyone but a studio that have specialized people and workflows for it.
So painful otherwise, and I've seen some things...(probably a reason why I'm so worn out from things like this.)
I want to use OT but I don' want to use OT anymore.

**Comment by shun-iwasawa** [comment]
Thank you @Lovewyrm for the suggestion.

> Why not freeze OT as it is and make something new?

For me, it is mainly because that OT is currently used in our studio, and it works fine so far.
As long as they produce the film with it, there are demands to tackle with.
And usually it is difficult for them to migrate their production software during the project.

If you love to make a new software rather than to amend the existing code (I think it would be always YES for most of developers) then you can do it without freezing OT. I think our OSS license is lax enough to make a new software based on it.

Anyway, any improvements to OT will be welcome. Making OpenGL to modern will be very very much welcome.

**Comment by Lovewyrm** [comment]
Well, if OT works fine for your studio, then why should the public continue working on it?
Are the issues only on our side? Doesn't it crash for you? Is using OT a fun experience for you?
Would you go hungry if you didn't use it? Are you stockholmed by it? Do you not at all loathe its quirks? 

I don't think many people outside of big(ger) studios are benefiting at all from ensnaring developers to OT.
Sure, no need to freeze OT, but then people who could work on something new continue to tame this surly monkey of a software package. Gotta treat it just right. Gotta bend the knee to its quirks.

Well, maybe if I had some assistants that deal with some of the cruft for me, but, alas, I don't have them.
The thing doesn't even work right on my OS, on the tablet even with patches to Qt, that fixes one thing and then breaks another (right now I can do shift and trace due to some patches, without losing F1, F2, and F3 focus, but that broke color model picking, as in, picking from the image/squares in the color model window, it works but it's very broken and requires multiple taps inside and outside of its tooldocker)

What's in for us? 
How does OT give any of us wings.

Us small scalers, who do this with little or no funding and with little or no assistants, much less specialized departments.

Man even thinking about this cheeses me off all over again.
In hindsight I wish Toonz never went Open, now it's a roach motel that benefits pretty much noone.

**Comment by manongjohn** [comment]
While I agree in many respects that OT isn't necessarily non-studio user friendly, it has come a long way since it's debut so I think it's working it's way their slowly. I don't think it's as bad as you feel it is, but I also know it still has a number of issues that have largely gone unaddressed that keep it from excelling.  We all know that OT isn't the greatest, nor does it fit into everyone's creative processes and workflows.  It is what it is a free, opensource animation tool.

As it's a free tool and opensource, you deal with what is there unless you are willing to make changes to it, either directly or indirectly. Nothing is stopping anyone from completely branching OT into their own thing, the problem is no one is interested in doing that, don't have the time or don't have people and funds to do it.  It's difficult enough keeping the few OT devs we've had that have contributed to keep contributing and improving it without any real incentive. (Guess I'm the odd ball). Comments like yours don't help it any.

At any rate, if OT doesn't work for you, then don't use it...move on. No one is forcing you to stay with OT. But thank you for your opinion, nonetheless.  I'm sure there are others that feel as you do, but they've probably moved on instead of posting insulting comments about it here where people are actually working on trying to improve what we have been given.

**Comment by Lovewyrm** [comment]
I know how free software works.
I'm concerned about the time and effort versus actual improvement and sunk cost fallacies.

Do you think that the slow walk toward a 'someday really good version' is faster and yields better results than cutting losses and making it from scratch instead of having to dig through its guts?

Also, "don't use it if it's so bad" is not THAT good of an argument to keep working on it.
I mean, you're right. Don't use it if you don't like it, but it's not black and white like that.

People try to make things work, usually, anyway. Especially people who even agree to deal with free software. 
Meaning, they don't drop things immediately, usually.

Meaning the crap that is within OT will affect people to some extent until they finally break.
I'm an example of it, a few months ago my opinion was completely different, now I want OT to not even exist anymore.
Why would I be like this? For fun? 
Maybe you just have a different breaking point. I have reached mine. 

Anyway, the important question, once more:
Do you think that the slow walk toward a 'someday really good version' is faster and yields better results than cutting losses and making it from scratch instead of having to dig through its guts?

**Comment by gab3d** [comment]
@Lovewyrm 
It's all about resources.
I'm sure there's no point in starting a new software from zero, but instead it's possible to build on what we have, and improve what is already there and should be improved.
Given enough developers working on it I'm sure all of what has been tormenting you would be gone.

So to get down to the ground, I'd like to know a couple of details:
- What are your most offending 5 problems with OT?
- What are you using it for, and in which environment?
- How much funds/time/developing efforts would you be willing to contribute monthly to the project to solve those problems?

If, at the end, you still think a new software will have a better result you should check **[enve](https://www.youtube.com/channel/UCjvNO3xdIQs-cpgwSf0mX5Q/videos)**, a new animation program that is at its infancy, but has all the right elements in place to be a good software in the future. 
While you see it, please try to visualize how much time would it take to cover all the areas that OT covers with a certain level of proficiency, a relevant user base, multi-platform, and be used in serious productions as a way of testing its strength, etc. 

For our developers: **enve** is a great proof of concept, with a lot of things that would be godsent in OT. Please take a good look at it for inspiration. 👍

**Comment by manongjohn** [comment]
I'm sorry, but do you see a ton of developers actively working on OT right now? Change can only happen so fast and you need developers for that, so if you truly think insulting OT is the way to get support, good luck.

OT has potential, in my opinion, but it lacks enough developers to achieve that.  Not only that, you need developers who are willing to work on stuff they may not really be interested in and can tackle some of the more intricate logic that needs to be redone. We also have no real organization and leadership as a result.  Devs come and go as they please and work only on what they want or have been paid by someone else to do. 

The only way you are going to see what you want is to hire and fund a full time developer to do it.  Got lots money?

I'd ask you why the sudden disdain for OT, but I'd probably just get more insulting comments rather than constructive dialogue for improvement.

**Comment by artisteacher** [comment]
There's nothing stopping someone from using OpenToonz as a starting point or inspiration for something new. Serif's Affinity suite is a great example of the benefits of starting over a codebase from scratch. For Serif, their new suite took years of developing. Clearly they had the funding to make that feasible. If anyone has the money/drive/time to make it happen - totally go for it. I don't see why OT should have to freeze for that to happen.  Starting over would take a really long time to get something with the range of capabilities we currently have in OT.

Anyways, I really appreciate the people who continue to give their time and energy to improving OT. In my experience, there's room for growth, but it has really improved a lot over the last couple years.

**Comment by jrhyme** [comment]
Author, no offense in advance
I remember a similar troll already encountered in the group of the website vk.kom, where the lessons on opentoonz, 
which wrote that in 2 years the program did not bring anything new and from the innovations - became open ... 
laughed for a long time with Kostya from morevna. The post of the author of the topic resembles a child who writes - the application does not start and puts one star.
Use your flash and toon boom and learn what pain is. Especially in character rigging. To make even a primitive mask of the eyeball, it takes many times less time than in the same toon boom.
It is necessary not to whine but to study.

**Comment by beeheemooth** [comment]
1. OpenGL
2. Vectors
3. IK
4. UI
5. History

**Comment by Lovewyrm** [comment]
@manongjohn 
I don't know what you mean by "more insulting comments".
Is it me reacting negatively to @shun-iwasawa saying that the software is working fine for his studio?

Well, that is just confusion on my part, because the software is not working fine for many (hence all the issues)
Are we all using some sort of inferior outsider version that none of us could use in production unless we risk:
Loss of files, either because the software is flat out deleting them, or not saving them.
Loss of progress like vector fills going in and out depending on fill gap values (unless a very strict regiment of OT desired vector approach is used) and other such quirks.
Great confusion at interpolation artifacts with raster animation when the animation tool is used, which uses non whole number values for pixels, causing OT to smear them and the output to look wrong.
And other such things pertaining to actual production, nevermind all the general instabilities that are not there in Mr. Iwasawas studio since it's working fine there.
That's my problem, that "oh it's working just fine". Why doesn't it work fine for me? 
What would I have to do to make it work fine for me? 

I am a programmer, too. I did read into the OT code but I've seen flat out comments in it that laughed at how hacked together some parts are. (Although, truth be told I have forgotten where that was exactly, I think it was pertaining to some OGL stuff, either it directly or making it play along with Qt).

My disdain is not that sudden, it's the kind of disdain that grew over time, with many little cuts from using it.
I don't want to attempt to retell them all, for my blood pressures sake. Even if that risks me coming across as "c'mon this guy has nothing to say, he's pulling it all out of his rear", I'd rather have people think that than seeing some stockholmed guy do one of those replies where every point I make gets quoted one by one with some explanation how it's not a problem for them and I'm just talking out of my rear, really, and I have to cowboy up. 
(Then you will get something like my response to Mr. Iwasawa )

I'm not a troll, though, I even feel bad about this, and I get the angle of just shutting up, but I really, really doubt that private people get anything good from this. Even people who are 'in the know' now and do many a tutorial often step around issues and comment on that. 


Anyway, I think I have a solution, or at least the starts of it, something that could cut costs both in time and money dramatically because it tackles the core of the issue of animation.

Maybe, maybe what the world needs is...a standard, a standard protocol that media programs implement, that allows inter program/process time syncing, and asset manipulation via keyframes.

That way programs could be developed independently, as long as they have an interface for keyframing their abilities. Something like MIDI for media sync.

I came up with this in bed so it's probably rather half baked and I certainly have to mull over it more, but I envisioned a standalone XSheet, graph editor and what's it called again, transporter? Those strips with playback controls. Keeping other, external programs in sync, which react to events like "show this and that frame" with "object here and there" and "attribute set to this" and "layers blended like this", with that protocol.

It works for music production and DAWs... some of them being even completely detached and 'free floating' separate processes.
(NonDAW https://non.tuxfamily.org/ )

Anyway, I'm also looking at those other projects for now.
Closing this issue for the time being but I'll respond to further messages, of course.

**Comment by manongjohn** [comment]
> In hindsight I wish Toonz never went Open, now it's a roach motel that benefits pretty much noone.

I can't speak for anyone else, but as someone who has invested a lot of time trying to improve this software, I find the above comment you made extremely insulting to my efforts.  There are others comments you have made that I also find insulting but this one seems the worst.

Don't get me wrong. I understand your frustrations with OT.  I don't think it's lost on anyone here. We all know OT has issues but no one to fix them, especially the bigger issues.  It's worse for macOS users since we can't get macOS developers to fix apparent platform specific problems which seems to get worse with each new OS upgrade.

I've been saying for quite a while that some of our issues probably stem from the fact that we don't really use OpenGL right.  We had a hasty attempt at using OpenGL better but had reverted parts of it due to issues and never went back to address it properly. Instead we've just kept patching it, waiting for QT to solve some of them for us. Honestly, after that debacle, I wanted to ditch OT also, but I decided to stick it out anyway. I also believe there are issues with vector intersection calculations and management which impacts fills and causes random crashes.  Wish I could fix these but I don't have the knowledge or time to do it.

Not sure what is left to say. As a developer, if you feel starting over is better and you have time to do it, then please lead the way.  Let us know when you have something we can play with and help improve. I'd love to see what you come up with.

**Comment by Lovewyrm** [comment]
Ah, I see now why you feel insulted. I figured it would be more insulting to work for something that has very little benefit for private people.

From what I have gathered, Toonz worked well enough for all that professional accolade.
Then it turned open source, and it still works fine for studios it seems even though regular folk seem to have issues left and right.

So what did the general population gain? Yes it's technically true that right now, because of the license changes, I can use a piece of software that was at least part of really big productions.

...but after using it for some time I wonder how much good this does me, or others who are not part of big studios or don't happen to have some sort of prerequisite mindset to using the software right.

Yet regular folk work on it, I wonder if at this point in time, it's not more regular folk than 'studio folk' are working on it.

It's like working on some industrial piece of hardware if it had no direct resource cost and I could just summon some epic CNC milling machine into my backyard to carve me a pointy stick, only in order to use it I have to learn some old industrial standard formerly proprietary language, and you really just want a pointy stick but the UK rules the world now so you don't have any carving knives unless you're government elite who are allowed knives.

But even this analogy falls apart of course, but just think of a bunch of private people trying to work with this miraculous machine, spending hours and hours of their lives 'making it usable for the common man' and it's advertised as professional grade but most people just hobble along with it and produce things that are even more crude than a pointy stick, or something equally silly.

Its one of those white elephant punishment gifts to me now. Only, I can (and am) reject it.
Life was possible before opentoonz, but now it exists and it ensnared people with its useless professional label and fame.
I never meant to insult your efforts, I am angry that they are being poured into this piece of software with only little merit. (After all, toonz worked before OT, and in this state, as per Mr Iwasawas statement, also works just fine, so I also don't know why it's bad to freeze it to free up the developers. Maybe he's forced to talk like this, that's what I'm going for in my mind)

Of course that is also rude because I shouldn't speak for anyone but myself, but that is why this issue was and is a question at heart, even though I have to admit that my diplomatic range is rather limited on it.

I guess I can reword my issue a bit, into a non question this time.

Let industry professionals have their precious software. Let us go for something else.

**Comment by gab3d** [comment]
Frankly, I got tired of reading your mind mambo...
Let's put it simple: we all know the software has it's issues (like every other software out there, free or not), we all know why there is only a certain amount of progress over time and not more (lack of developing resources), we all know OT doesn't currently fit every possible animator or studio requirement, BUT, we also know the software does allow to achive a lot of things in a very good way, it allows to tackle not only small but also big productions with lots of artists involved, and provides ways for doing it with the least cost possible on licenses and the freedom to change the code if needed.

So... perhaps you're part of the people that don't really find benefit in using OT, then you are absolutely free to go find a better solution for producing what you want. There are several excellent products on the market that you could benefit from.

But, I don't see the point in all of what you're saying... really!
If you want to do some kind of catharsis derived from the fact that the world is not like you wanted to be, I'd say you're in the wrong place.
If you really want to improve OpenToonz (and that's what this site is for) then you can contribute to it: 
- concrete and specific ideas to improve some feature or fix a bug.
- improve some feature or fix a bug yourself, since you are a developer.
- pay or help pay for someone to fix the bugs that you wish solved, or the features you want improved.

If not, I don't see the point in writing things here. 🤷‍♂️

Luck!

**Comment by manongjohn** [comment]
I can assure you I am not "studio folk" developing OT :)

Watching and listening to other animators that use OT, many non-professionals, I have seen their struggles and issues and I have tried to fix them where I can. My focus has never been what the studio wants/needs. All I want is for newbie/hobby animators, like me, to have access to OT's many features.  I am trying to make that easier for us to learn and use. 

Sometimes my changes are not agreeable as they might impact existing studio usage.  It's rather frustrating when you develop something and told to change it or make it an option because the studio may not like it even though many other users complain about it. It's happened more than once, but I made the change to make it acceptable to professional and non-professionals alike.  Better than nothing.

When I opened a discussion about a non-studio focused version to give developers more freedom in what can be changed and what to put into OT, it was kind of shut down.  Could I have branched off anyway? Sure, but without any reliable support and resources, that was going to be difficult for me so I didn't pursue it. 

At the end of the day, I don't feel I've wasted my time developing or learning OT.  If you feel you've wasted your time, then I am sorry you feel that way. Can't always tell something is right for you until you've worked with it for a bit.

Good luck in your endeavors.

**Comment by Lovewyrm** [comment]
@gab3d 
My point is simple.
I understand everything about free software, that's why I asked if it's worth continuing.
The answer seems to be "yes". I think it's misplaced time because, no, I absolutely do not agree that opentoonz is at all meant for general usage, the software code base as it is, and the application itself that emerges from it, is actively hostile toward usage that does not fall into it's intended usage.

Like OpenGL.
It's actually not a problem at all for OT, because in a studio, you just throw more hardware at it, if that were not the truth then toonz would not have enabled the production of the things that were produced with it. OT is production ready. It has made things before, and it's "working fine" right now. It does not need any further development. It made Ghibli productions way before this time and right now it's working in another production. It cares not a BIT about plebeian performance woes because a study has the hardware to cope with it.

That said, I have no doubt that, with enough effort, OT can someday be usable for everyone.
But.
BUT.

I don't see that happening faster than writing something like it from scratch since you are struggling against the very core of toonz.
If toonz could feel emotion then it would feel hate toward everything that isn't its original purpose, and it would feel loathing every time someone tries to tame it.

You can of course work all your life for it, and maybe OT will be the only thing standing, I merely doubt it and nobody who used OT and failed because of its non intended use case that private/small scale projects are will never get that time back.

Anyway, I've been a downer for enough. I'm moving on.
I'd wish you all luck back, but you don't need luck for this, only more work on it.

Good health to you all though.

**Comment by esotericDisciple** [comment]
Hi, Opentoonz MUST be maintained and developed as it's the ONLY linux 2d Animation app with every desiderata, ESPECIALLY thumbnail-preview timelines, AND audio timeline support...No other free animation software has both and worse none of them have an interest in adding them...it's as if Linux developers don't believe computers should be used by artists at all...
